python/ray/data/_internal/execution/autoscaler/default_autoscaler.py

```diff
@@ -52,73 +59,60 @@ def _calculate_actor_pool_util(self, actor_pool: AutoscalingActorPool):
         else:
             return actor_pool.num_active_actors() / actor_pool.current_size()
 
-    def _actor_pool_should_scale_up(
+    def _derive_scaling_action(
         self,
         actor_pool: AutoscalingActorPool,
         op: "PhysicalOperator",
         op_state: "OpState",
-    ):
+    ) -> _AutoscalingAction:
         # Do not scale up, if the op is completed or no more inputs are coming.
-        if op.completed() or (op_state.total_enqueued_input_bundles() == 0):
-            return False
+        if op.completed() or (
+            op._inputs_complete and op_state.total_enqueued_input_bundles() == 0
+        ):
+            return _AutoscalingAction.SCALE_DOWN
+
         if actor_pool.current_size() < actor_pool.min_size():
             # Scale up, if the actor pool is below min size.
-            return True
-        elif actor_pool.current_size() >= actor_pool.max_size():
+            return _AutoscalingAction.SCALE_UP
+        elif actor_pool.current_size() > actor_pool.max_size():
             # Do not scale up, if the actor pool is already at max size.
-            return False
+            return _AutoscalingAction.SCALE_DOWN
+
         # Do not scale up, if the op does not have more resources.
         if not op_state._scheduling_status.under_resource_limits:
-            return False
+            return _AutoscalingAction.NO_OP
+
         # Do not scale up, if the op has enough free slots for the existing inputs.
         if op_state.total_enqueued_input_bundles() <= actor_pool.num_free_task_slots():
-            return False
-        # Determine whether to scale up based on the actor pool utilization.
-        util = self._calculate_actor_pool_util(actor_pool)
-        return util > self._actor_pool_scaling_up_threshold
+            return _AutoscalingAction.NO_OP
 
-    def _actor_pool_should_scale_down(
-        self,
-        actor_pool: AutoscalingActorPool,
-        op: "PhysicalOperator",
-        op_state: "OpState",
-    ):
-        # Scale down, if the op is completed or no more inputs are coming.
-        if op.completed() or (op_state.total_enqueued_input_bundles() == 0):
-            return True
-        if actor_pool.current_size() > actor_pool.max_size():
-            # Scale down, if the actor pool is above max size.
-            return True
-        elif actor_pool.current_size() <= actor_pool.min_size():
-            # Do not scale down, if the actor pool is already at min size.
-            return False
-        # Determine whether to scale down based on the actor pool utilization.
+        # Determine whether to scale up based on the actor pool utilization.
         util = self._calculate_actor_pool_util(actor_pool)
-        return util < self._actor_pool_scaling_down_threshold
+        if util >= self._actor_pool_scaling_up_threshold:
+            return _AutoscalingAction.SCALE_UP
```

Comment:
In the current implementation, we don't scale up if the previous scale up hasn't finished. I think this diff might remove that behavior?
https://github.com/anyscale/rayturbo/blob/e156a9f9424ec1e499e93892d1a33d4bfb8599a1/python/ray/anyscale/data/autoscaler/anyscale_autoscaler.py#L357-L360